File: 312-burst-balloons/burst-balloons.cpp
class Solution {
public:
    int maxCoins(vector<int>& nums) {
        int n = nums.size();

        vector<int> balloons;
        balloons.push_back(1);
        for (int i = 0; i < n; i++)
            balloons.push_back(nums[i]);
        balloons.push_back(1);

        vector<vector<int>> dp(n + 2, vector<int>(n + 2, 0));

        for (int i = n + 1; i >= 1; i--) {
            for (int j = i + 1; j < n + 2; j++) {
                int maxCoins = INT_MIN;
                for (int k = i; k <= j - 1; k++) {
                    int coins = balloons[i - 1] * balloons[k] * balloons[j] +
                                dp[i][k] + dp[k + 1][j];
                    maxCoins = max<int>(maxCoins, coins);
                }

                dp[i][j] = maxCoins;
            }
        }

        return dp[1][n + 1];
    }
};

File: 312-burst-balloons/burst-balloons.java
// class Solution {
//     List<List<Integer>> mem;
//     private int solve(int[] arr, int i, int j) {
//         if (i == j) {
//             mem.get(i).set(j, 0);
//             return 0;
//         }
//         if (mem.get(i).get(j) != -1) {
//             return mem.get(i).get(j);
//         }

//         int maxi = Integer.MIN_VALUE;
//         for (int k = i; k <= j - 1; k++) {
//             int cost = arr[i - 1] * arr[k] * arr[j] +
//                        solve(arr, i, k) +
//                        solve(arr, k + 1, j);
//             maxi = Math.max(maxi, cost);
//         } 

//         mem.get(i).set(j, maxi);
//         return maxi;
//     }
//     public int maxCoins(int[] nums) {
//         int n = nums.length;
//         mem = new ArrayList<>();
//         for (int i = 0; i < n + 2; i++) {
//             mem.add(new ArrayList<>());    
//             for (int j = 0; j < n + 2; j++) {
//                 mem.get(i).add(Integer.valueOf(-1));
//             }
//         }

//         int[] arr = new int[n + 2];
//         arr[0] = 1;
//         arr[n + 1] = 1;
//         for (int i = 1; i <= n; i++) {
//             arr[i] = nums[i - 1];
//         }

//         return solve(arr, 1, n + 1);
//     }
// }

class Solution {
    int[][] mem;
    private int solve(int[] arr, int i, int j) {
        if (i == j) {
            return mem[i][j] = 0;
        }
        if (mem[i][j] != -1) {
            return mem[i][j];
        }

        int maxCoins = Integer.MIN_VALUE;
        for (int k = i; k <= j - 1; k++) {
            int coins = arr[i - 1] * arr[k] * arr[j] +
                       solve(arr, i, k) +
                       solve(arr, k + 1, j);
            maxCoins = Math.max(maxCoins, coins);
        }

        return mem[i][j] = maxCoins;
    }
    public int maxCoins(int[] nums) {
        int n = nums.length;
        mem = new int[n + 2][n + 2];
        for (int i = 0; i < n + 2; i++) {
            Arrays.fill(mem[i], -1);
        }

        int[] arr = new int[n + 2];
        arr[0] = 1; arr[n + 1] = 1;
        System.arraycopy(nums, 0, arr, 1, n);

        return solve(arr, 1, n + 1);
    }
}